lecture-notes/2-vcs.md

```diff
@@ -420,11 +475,607 @@ commits; the general write-add-commit flow; that all Git objects are stored in
+and asks you whether or not to stage it[^add-p-editing]. As usual, a subsequent
+`git commit` will include only the staged hunks, letting you quickly and easily
+distribute your changes across multiple commits.
+
+[^add-p-editing]: You can also choose to edit a hunk before staging it, which
+    will cause `git add -p` to open a text editor where you can modify or delete
+    lines. This isn't typically necessary, but it helps when unrelated changes
+    are close enough together that Git counts them as a single hunk. (For
+    changes that are close but not directly adjacent, you can also ask `git add
+    -p` split the hunk into multiple smaller hunks and prompt for each.)
+
+```
+TODO: Example of `git add -p` to stage specific hunks that have changed.
+Demonstrate committing multiple changes separately with good messages.
+```
+
+### Bringing it all together
+
+With `git add -p` and some commit guidelines under your belt, you have
+everything you need to start tracking a project's history in Git. Here's how
+that might look for a simple project (which will look familiar once you begin
+Homework 4!). First, you create a repository:
+
+```console
+$ mkdir calc
+$ cd calc
+$ git init
+Initialized empty Git repository in /home/you/calc/.git/
+```
+
+Then, you commit some initial code:
+```console
+$ cat main.c
+int main() {
+  printf("Hello, world!\n");
+  return 0;
+}
+$ git add main.c
+$ git commit -m "Initial commit"
+[main (root-commit) 3ba2a93] Initial commit
+ 1 file changed, 3 insertions(+)
+ create mode 100644 main.c
+```
+
+You fix mistakes in the code...
+```console
+$ git diff
+diff --git a/main.c b/main.c
+index fbd71ab..aed773b 100644
+--- a/main.c
++++ b/main.c
+@@ -1,3 +1,5 @@
++#include <stdio.h>
++
+ int main() {
+   printf("Hello, world!\n");
+   return 0;
+
+$ git add main.c
+$ git commit -m "Add #include directives to fix compilation"
+[main a83e7a6] Add #include directives to fix compilation
+ 1 file changed, 2 insertions(+)
+```
+
+...continue working on the project...
+```console
+$ cat main.c
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+int calc(int left, char op, int right) {
+  <some code>
+}
+
+int main(int argc, char **argv) {
+  if (argc != 4) {
+    fprintf(stderr,
+            "Usage: %s <num> <op> <num>\nWhere <op> is one of +, -, *, /.\n",
+            argv[0]);
+    return EXIT_FAILURE;
+  }
+  const char *left_str = argv[1];
+  const char *op_str = argv[2];
+  const char *right_str = argv[3];
+  errno = 0;
+  int left = strtol(left_str, NULL, 10);
+  if (errno != 0) {
+    perror(argv[0]);
+    return EXIT_FAILURE;
+  }
+
+  <some more code>
+
+  int result = calc(left, op_str[0], right);
+  fprintf(stdout, "%d\n", result);
+  return 0;
+}
+$ git add main.c
+$ git commit -m "Make main a simple calculator"
+[main 1838eea] Make main a simple calculator
+ 1 file changed, 45 insertions(+), 2 deletions(-)
+```
+
+...make some smaller edits...
+```console
```

Comment:
rouge does `console?lang=python,prompt=>>>,...` so perhaps you could do `console?lang=diff` or instead just do a normal git diff with `diff` highlighting and explain what command you got to get it

Reply:
or maybe it's `console?lang=python&prompt=>>>,...`

Reply:
"and ?comments to enable # comments"